### DISCUSSION TITLE
Bind window maximization request

### DISCUSSION THREAD

**Discussion by megatux** [YaLTeR/niri#293] [Ideas]
Nowadays apps have client side window decorations and it's common to have the maximize button. It'd be nice to be able to bind that app event to a compositor action like `maximize-column`. Is it trappable?

**Comment by YaLTeR** [comment]
The problem is that this xdg-shell maximize implies some behavioral differences. The surface must be set to a maximized state, which makes the corners square instead of round (similar to the tiled state of `prefer-no-csd`) but also requires the applications to match the requested geometry exactly. We have to set the maximized state, because "maximize" and "unmaximize" are separate requests and the application needs to know which one to send.

Essentially, I didn't want to have this separate code path with subtly different semantics.

**Reply by Supreeeme** [reply]
Does it really imply square corners? I don't see anything in the spec mandating how the corners behave.

The ability to handle maximization state would be useful for touch devices.

**Reply by YaLTeR** [reply]
How would it be useful for touch devices specifically?

**Reply by Supreeeme** [reply]
Right now it's hard to control window size using touch only out of the box. Adding the ability for clients to maximize themselves would provide an easy way to control window size using touch only for clients that provide a maximize button in their decorations.

I also think it's just generally a little surprising that the maximize-window action and the maximize button on decorations don't do the same thing.

**Reply by YaLTeR** [reply]
That makes sense, guess it's a bit more important for touch-only.

Either way, it would still be somewhat complex to implement. It should have semantics similar to the fullscreen state (expel into own column, expand to full working area size), but respecting the layer-shell exclusive area.

**Comment by YaLTeR** [comment]
I'm implementing proper Wayland maximize here: https://github.com/YaLTeR/niri/pull/2376
